---
permalink: ipython-notebooks/bioinformatics/
description: 5 popular visualizations that bioinformaticians use in exploratory analysis of genomic data.
title: Visualizing biological data&#58; exploratory bioinformatics with plot.ly
has_thumbnail: false
thumbnail: /images/static-image
layout: user-guide
name: Visualizing Biological Data
language: python
page_type: u-guide
---
{% raw %}
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>About the author:
Oxana is a data scientist based in Stockholm, Sweden. She is studying for a PhD in Bioinformatics, exploring molecular evolution patterns in eukaryotes. You can follow Oxana on Twitter <a href="http://twitter.com/Merenlin">@Merenlin</a> or read <a href="http://merenlin.com">her blog</a>.</p>
<h4 id="Introduction">Introduction<a class="anchor-link" href="#Introduction">&#182;</a></h4><p>This notebook will give you the recipes of the most popular data visualizations I encounter in my work as a bioinformatician. If you always wondered what bioinformatics is all about or would like to create interactive
visualization for your genomic data using <a href="https://plot.ly/python/">plot.ly</a>, this is the place to start.</p>
<p>We will be working with real <a href="http://en.wikipedia.org/wiki/Gene_expression">gene expression</a> data obtained by <a href="http://en.wikipedia.org/wiki/Cap_analysis_gene_expression">Cap Analysis of Gene Expression(CAGE)</a> from human samples by the <a href="http://fantom.gsc.riken.jp/5/">FANTOM5</a> consortium. We will be following a typical workflow of a bioinformatician exploring new data, looking for the outliers: interesting genes or samples, or general patterns in the data. In the end, you'll get the idea of the challenges and upsides of creating interactive visualizations of biological data using plot.ly Python API.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Obtaining-the-data">Obtaining the data<a class="anchor-link" href="#Obtaining-the-data">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>FANTOM5 provides high precision data of thousands of human and mouse samples. The vastness of this data can be overwhelming and operating it locally is challenging. Luckily, there are many tools out there to make our life easier.<br>
For creating a small data subset we can work with in this tutorial, I used <a href="http://fantom.gsc.riken.jp/5/tet">TET: Fantom 5 Table Extraction tool</a>. I picked a few human samples, mostly brain tissues with a few outliers, like uterus and downloaded a tab-separated file from the website. For more advanced data extraction, it's good to have a look at <a href="https://github.com/Hypercubed/TET/blob/master/README.md">TET's API</a>. 
I have picked normalized tpm(tags per million) and annotated data, so we can focus only on processed data for protein coding genes. All data files for this notebook are available on figshare: <a href="http://dx.doi.org/10.6084/m9.figshare.1430029">http://dx.doi.org/10.6084/m9.figshare.1430029</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Loading-the-dataset">Loading the dataset<a class="anchor-link" href="#Loading-the-dataset">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We are loading the data from the .tsv file, skipping the first two columns (00Annotation and short_description).</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[47]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="kn">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">genfromtxt</span><span class="p">(</span><span class="s2">&quot;http://figshare.com/download/file/2087487/1&quot;</span><span class="p">,</span>
                     <span class="n">comments</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">73</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">names</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">object</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
<span class="k">print</span> <span class="s2">&quot;Number of genes: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>Number of genes: 201802
</pre>
</div>
</div>

<div class="output_area">
<div class="prompt output_prompt">Out[47]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>uniprot_id</th>
      <th>Astrocyte__cerebellum_donor1CNhs1132111500119F6</th>
      <th>Astrocyte__cerebral_cortex_donor1CNhs1086411235116D2</th>
      <th>brain_adult_donor1CNhs1179610084102B3</th>
      <th>brain_adult_pool1CNhs1061710012101C3</th>
      <th>brain_fetal_pool1CNhs1179710085102B4</th>
      <th>breast_adult_donor1CNhs1179210080102A8</th>
      <th>cerebellum__adult_donor10196CNhs1379910173103C2</th>
      <th>cerebellum_adult_donor10252CNhs1232310166103B4</th>
      <th>cerebellum_newborn_donor10223CNhs1407510357105E6</th>
      <th>...</th>
      <th>thalamus__adult_donor10196CNhs1379410168103B6</th>
      <th>thalamus_adult_donor10252CNhs1231410154103A1</th>
      <th>thalamus_adult_donor10258_tech_rep1CNhs1422310370105G1</th>
      <th>thalamus_adult_donor10258_tech_rep2CNhs1455110370105G1</th>
      <th>thalamus_newborn_donor10223CNhs1408410366105F6</th>
      <th>throat_fetal_donor1CNhs1177010061101H7</th>
      <th>thyroid_fetal_donor1CNhs1176910060101H6</th>
      <th>tongue_epidermis_fungiform_papillae_donor1CNhs1346010288104F9</th>
      <th>umbilical_cord_fetal_donor1CNhs1176510057101H3</th>
      <th>uterus_fetal_donor1CNhs1176310055101H1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>             NA</td>
      <td>    0</td>
      <td>     0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>...</td>
      <td>   0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
    </tr>
    <tr>
      <th>1</th>
      <td> uniprot:Q96JB6</td>
      <td> 0.12</td>
      <td> 11.45</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 2.17</td>
      <td> 0</td>
      <td> 0.22</td>
      <td> 1.03</td>
      <td>...</td>
      <td> 5.8</td>
      <td> 0.31</td>
      <td> 5.65</td>
      <td> 2.99</td>
      <td> 0</td>
      <td> 1.19</td>
      <td> 1.01</td>
      <td> 2.58</td>
      <td> 7.04</td>
      <td> 4.48</td>
    </tr>
    <tr>
      <th>2</th>
      <td>             NA</td>
      <td>    0</td>
      <td>     0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>...</td>
      <td>   0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>             NA</td>
      <td>    0</td>
      <td>     0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>...</td>
      <td>   0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>             NA</td>
      <td>    0</td>
      <td>     0</td>
      <td> 0</td>
      <td> 0</td>
      <td> 0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>...</td>
      <td>   0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td> 0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
      <td>    0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 71 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Let's also make sure that we filter out those genes for which the <a href="http://www.uniprot.org/">Uniprot</a> Id is unknown. That will reduce our data, besides, we are only interested in proteins in this analysis.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[48]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="n">uniprot_clean</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;uniprot_id&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="p">(</span><span class="n">x</span> <span class="o">!=</span> <span class="s1">&#39;NA&#39;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">((</span><span class="n">x</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">))]</span>
<span class="n">df</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;uniprot_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">uniprot_clean</span><span class="p">)]</span>
<span class="k">print</span> <span class="s2">&quot;Number of genes: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">))</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt"></div>

<div class="output_subarea output_stream output_stdout output_text">
<pre>Number of genes: 59173
</pre>
</div>
</div>

<div class="output_area">
<div class="prompt output_prompt">Out[48]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<div style="max-height:1000px;max-width:1500px;overflow:auto;">
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>uniprot_id</th>
      <th>Astrocyte__cerebellum_donor1CNhs1132111500119F6</th>
      <th>Astrocyte__cerebral_cortex_donor1CNhs1086411235116D2</th>
      <th>brain_adult_donor1CNhs1179610084102B3</th>
      <th>brain_adult_pool1CNhs1061710012101C3</th>
      <th>brain_fetal_pool1CNhs1179710085102B4</th>
      <th>breast_adult_donor1CNhs1179210080102A8</th>
      <th>cerebellum__adult_donor10196CNhs1379910173103C2</th>
      <th>cerebellum_adult_donor10252CNhs1232310166103B4</th>
      <th>cerebellum_newborn_donor10223CNhs1407510357105E6</th>
      <th>...</th>
      <th>thalamus__adult_donor10196CNhs1379410168103B6</th>
      <th>thalamus_adult_donor10252CNhs1231410154103A1</th>
      <th>thalamus_adult_donor10258_tech_rep1CNhs1422310370105G1</th>
      <th>thalamus_adult_donor10258_tech_rep2CNhs1455110370105G1</th>
      <th>thalamus_newborn_donor10223CNhs1408410366105F6</th>
      <th>throat_fetal_donor1CNhs1177010061101H7</th>
      <th>thyroid_fetal_donor1CNhs1176910060101H6</th>
      <th>tongue_epidermis_fungiform_papillae_donor1CNhs1346010288104F9</th>
      <th>umbilical_cord_fetal_donor1CNhs1176510057101H3</th>
      <th>uterus_fetal_donor1CNhs1176310055101H1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1 </th>
      <td>                               uniprot:Q96JB6</td>
      <td>  0.12</td>
      <td> 11.45</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
      <td> 2.17</td>
      <td>     0</td>
      <td>  0.22</td>
      <td>  1.03</td>
      <td>...</td>
      <td>   5.8</td>
      <td>  0.31</td>
      <td> 5.65</td>
      <td>  2.99</td>
      <td>     0</td>
      <td>  1.19</td>
      <td>  1.01</td>
      <td> 2.58</td>
      <td>  7.04</td>
      <td>  4.48</td>
    </tr>
    <tr>
      <th>6 </th>
      <td>                               uniprot:Q8N2H3</td>
      <td>  7.51</td>
      <td>   6.3</td>
      <td>  3.88</td>
      <td>  3.71</td>
      <td>     2</td>
      <td> 5.07</td>
      <td>  1.53</td>
      <td>  1.99</td>
      <td>  6.72</td>
      <td>...</td>
      <td>  4.15</td>
      <td>  7.34</td>
      <td> 4.23</td>
      <td>  3.84</td>
      <td>  4.28</td>
      <td> 15.24</td>
      <td> 11.92</td>
      <td> 7.74</td>
      <td>  7.04</td>
      <td> 15.87</td>
    </tr>
    <tr>
      <th>7 </th>
      <td>                               uniprot:Q8N2H3</td>
      <td>  3.69</td>
      <td>  3.43</td>
      <td>  1.94</td>
      <td>  0.65</td>
      <td>     0</td>
      <td>    0</td>
      <td>  1.53</td>
      <td>  0.55</td>
      <td>  1.55</td>
      <td>...</td>
      <td>  0.83</td>
      <td>  2.69</td>
      <td> 2.82</td>
      <td>  1.92</td>
      <td>     0</td>
      <td>  3.33</td>
      <td>  3.03</td>
      <td> 2.58</td>
      <td>  2.35</td>
      <td>  5.29</td>
    </tr>
    <tr>
      <th>14</th>
      <td> uniprot:Q92902,uniprot:Q658M9,uniprot:Q8WXE5</td>
      <td> 35.58</td>
      <td> 20.61</td>
      <td> 30.05</td>
      <td> 21.71</td>
      <td> 19.96</td>
      <td> 31.9</td>
      <td> 13.73</td>
      <td> 20.79</td>
      <td> 21.72</td>
      <td>...</td>
      <td> 15.75</td>
      <td> 26.89</td>
      <td>   24</td>
      <td> 26.24</td>
      <td> 12.83</td>
      <td> 19.76</td>
      <td> 22.03</td>
      <td> 12.9</td>
      <td> 32.85</td>
      <td> 17.09</td>
    </tr>
    <tr>
      <th>23</th>
      <td>                               uniprot:Q8WWQ2</td>
      <td>     0</td>
      <td>     0</td>
      <td>     0</td>
      <td>  1.02</td>
      <td>     0</td>
      <td>    0</td>
      <td>     0</td>
      <td>  2.99</td>
      <td>     0</td>
      <td>...</td>
      <td>   5.8</td>
      <td>  8.38</td>
      <td> 4.23</td>
      <td>  8.96</td>
      <td>  0.71</td>
      <td>  0.71</td>
      <td>     0</td>
      <td>    0</td>
      <td>  7.04</td>
      <td>   5.7</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 71 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="1.-MA-scatter-plot-comparing-newborn-and-adult-tissues">1. MA scatter plot comparing newborn and adult tissues<a class="anchor-link" href="#1.-MA-scatter-plot-comparing-newborn-and-adult-tissues">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p><a href="http://en.wikipedia.org/wiki/MA_plot">MA plot</a> is a popular visualization tool coming from the microarray analysis. It allows researchers to explore true statistical differences between the two samples, arrays or other observations. We are going to look at the two samples of substantia nigra tissues from the brain of an adult and a newborn person. How do their genetic profiles differ?</p>
<p>Firs, let's subset our big dataframe to only include the samples of interest. We will also prefilter the data to not include genes that are not expressed in these tissues.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[49]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="n">df_MA</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s2">&quot;uniprot_id&quot;</span><span class="p">,</span><span class="s1">&#39;substantia_nigra_adult_donor10258CNhs1422410371105G2&#39;</span><span class="p">,</span><span class="s1">&#39;substantia_nigra_newborn_donor10223CNhs1407610358105E7&#39;</span><span class="p">]]</span>
<span class="n">df_MA</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span><span class="s1">&#39;adult&#39;</span><span class="p">,</span> <span class="s1">&#39;newborn&#39;</span><span class="p">]</span>
<span class="n">df_MA</span><span class="p">[[</span><span class="s1">&#39;adult&#39;</span><span class="p">,</span><span class="s1">&#39;newborn&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">df_MA</span><span class="p">[[</span><span class="s1">&#39;adult&#39;</span><span class="p">,</span> <span class="s1">&#39;newborn&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span> 
<span class="n">df_MA</span> <span class="o">=</span> <span class="n">df_MA</span><span class="p">[(</span><span class="n">df_MA</span><span class="o">.</span><span class="n">T</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">()]</span>  <span class="c1">#remove rows with all zeros</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>There are many different methods of computing the average expression level(A) between the two observations and
the mean variation(M). To keep things simple, for this example we will just compare the sum on the x-axis vs the minus on the y-axis.
Our data is already normalized and preprocessed, so this will be enough to find the clear outliers.</p>
<p>Here lies the firs problem with web-based interactive visualizations. Plot.ly at the moment has a very hard time rendering more than 40k points in a scatter plot. So for the sake of this example, I'll plot a subset of the data.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[50]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="kn">import</span> <span class="nn">plotly.plotly</span> <span class="kn">as</span> <span class="nn">py</span>
<span class="kn">from</span> <span class="nn">plotly.graph_objs</span> <span class="kn">import</span> <span class="o">*</span>

<span class="n">A</span> <span class="o">=</span> <span class="n">df_MA</span><span class="p">[</span><span class="s1">&#39;adult&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df_MA</span><span class="p">[</span><span class="s1">&#39;newborn&#39;</span><span class="p">]</span>
<span class="n">M</span> <span class="o">=</span> <span class="n">df_MA</span><span class="p">[</span><span class="s1">&#39;adult&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">df_MA</span><span class="p">[</span><span class="s1">&#39;newborn&#39;</span><span class="p">]</span>

<span class="n">trace</span> <span class="o">=</span> <span class="n">Scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">A</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">],</span>
    <span class="n">y</span><span class="o">=</span><span class="n">M</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">],</span>
    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers&#39;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;substantia nigra&quot;</span><span class="p">,</span>
    <span class="n">text</span><span class="o">=</span><span class="n">df_MA</span><span class="p">[</span><span class="s1">&#39;gene&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">],</span>
    <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="n">Line</span><span class="p">(</span>
            <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
        <span class="n">opacity</span><span class="o">=</span><span class="mf">0.8</span><span class="p">))</span>

<span class="n">layout</span> <span class="o">=</span> <span class="n">Layout</span><span class="p">(</span><span class="n">showlegend</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
                <span class="n">title</span><span class="o">=</span><span class="s2">&quot;MA plot of gene expression in adult and newborn samples of substantia nigra&quot;</span><span class="p">,</span>
                <span class="n">xaxis</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span>
                    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;A&#39;</span><span class="p">,</span>
                <span class="p">),</span>
               <span class="n">yaxis</span><span class="o">=</span><span class="n">YAxis</span><span class="p">(</span>
                    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;M&#39;</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Data</span><span class="p">([</span><span class="n">trace</span><span class="p">]),</span> <span class="n">layout</span><span class="o">=</span><span class="n">layout</span><span class="p">)</span>
<span class="n">py</span><span class="o">.</span><span class="n">iplot</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt output_prompt">Out[50]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<iframe id="igraph" scrolling="no" style="border:none;"seamless="seamless" src="https://plot.ly/~oxana/253.embed" height="525" width="100%"></iframe>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now we can already start exploring some of the genes, that behave differently in adult vs newborn samples.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="2.-Histograms-of-expression-breadth-and-average-expression-levels">2. Histograms of expression breadth and average expression levels<a class="anchor-link" href="#2.-Histograms-of-expression-breadth-and-average-expression-levels">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Another timeless visualization for exploratory data analysis is histogram. Here we don't need to subset our data anymore, for plots like these Plot.ly's capacity is up to 100k points. Let's see how expression breadth(in how many tissues the gene is expressed) and average expression levels look like for all of the samples.</p>
<p>Just use the "domain" variable to regulate where the axis of each subplot are.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[51]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;breadth&#39;</span><span class="p">]</span> <span class="o">=</span>  <span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;float&#39;</span><span class="p">)</span>

<span class="o">&gt;</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;float&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">trace1</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;expression breadth&quot;</span><span class="p">,</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;breadth&#39;</span><span class="p">],</span>
        <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span>
        <span class="n">line</span><span class="o">=</span><span class="n">Line</span><span class="p">(</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">0</span>
        <span class="p">),</span>
        <span class="n">opacity</span><span class="o">=</span><span class="mf">0.75</span>
        <span class="p">),</span>
<span class="p">)</span>

<span class="n">trace2</span> <span class="o">=</span> <span class="n">Histogram</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;average expression&quot;</span><span class="p">,</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg&#39;</span><span class="p">],</span>
        <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span>
        <span class="n">line</span><span class="o">=</span><span class="n">Line</span><span class="p">(</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mi">0</span>
        <span class="p">),</span>
        <span class="n">opacity</span><span class="o">=</span><span class="mf">0.75</span>
        <span class="p">),</span>
    <span class="n">xaxis</span><span class="o">=</span><span class="s1">&#39;x2&#39;</span><span class="p">,</span>
    <span class="n">yaxis</span><span class="o">=</span><span class="s1">&#39;y2&#39;</span>
    <span class="p">)</span>


<span class="n">layout</span> <span class="o">=</span> <span class="n">Layout</span><span class="p">(</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Exploring the distributions&quot;</span><span class="p">,</span>
    <span class="n">xaxis</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="s1">&#39;breadth&#39;</span><span class="p">,</span>
         <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.45</span><span class="p">]</span>
    <span class="p">),</span>
    <span class="n">xaxis2</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="s1">&#39;average expression&#39;</span><span class="p">,</span>
        <span class="n">domain</span><span class="o">=</span><span class="p">[</span><span class="mf">0.55</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="p">),</span> 
    <span class="n">yaxis2</span><span class="o">=</span><span class="n">YAxis</span><span class="p">(</span>
        <span class="n">anchor</span><span class="o">=</span><span class="s1">&#39;x2&#39;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Data</span><span class="p">([</span><span class="n">trace1</span><span class="p">,</span> <span class="n">trace2</span><span class="p">]),</span> <span class="n">layout</span><span class="o">=</span><span class="n">layout</span><span class="p">)</span>
<span class="n">py</span><span class="o">.</span><span class="n">iplot</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt output_prompt">Out[51]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<iframe id="igraph" scrolling="no" style="border:none;"seamless="seamless" src="https://plot.ly/~oxana/254.embed" height="525" width="100%"></iframe>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Here is where interactive visualization comes in handy. Average expression level distribution looks very wide because of a few outliers, - highly expressed genes and most of the genes actually being expressed at a very low level. But instead of trying to adjust the limits on the x-axis, we can just zoom in on the interesting area. Try it!</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="3.-Scatter-plot-with-a-trend-line">3. Scatter plot with a trend line<a class="anchor-link" href="#3.-Scatter-plot-with-a-trend-line">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This kind of plot must be the most popular way to visualize a trend in biological data. We seek clear
and simple patterns demonstrating the relationships between different biological parameters or observations.
Plot.ly's Python API does not come with out-of-the-box tools for plotting trend lines, but numpy has all we need.</p>
<p>Let's say we want to plot the relationship between the breadth of expression and the average level. Again, for speed and simplicity, we only take the first 1000 genes in our data frame. Let's try to fit a polinomial function to our data points and plot both at the same time. By using plot.ly it's simple, just send the regression line trace to the same figure.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[52]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;breadth&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;avg&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">]</span>
<span class="n">coefficients</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">polyfit</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="n">polynomial</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">poly1d</span><span class="p">(</span><span class="n">coefficients</span><span class="p">)</span>
<span class="n">r_x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">72</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">)</span>
<span class="n">r_y</span> <span class="o">=</span> <span class="n">polynomial</span><span class="p">(</span><span class="n">r_x</span><span class="p">)</span>

<span class="n">trace1</span> <span class="o">=</span> <span class="n">Scatter</span><span class="p">(</span>
    <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span>
    <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span>
    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers&#39;</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;expression levels&quot;</span><span class="p">,</span>
    <span class="n">text</span><span class="o">=</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;uniprot_id&#39;</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="mi">1000</span><span class="p">],</span>
    <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="n">Line</span><span class="p">(</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;rgba(217, 217, 217, 0.14)&#39;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
        <span class="n">opacity</span><span class="o">=</span><span class="mf">0.2</span><span class="p">))</span>

<span class="n">trace2</span> <span class="o">=</span> <span class="n">Scatter</span><span class="p">(</span>
    <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines+markers&#39;</span><span class="p">,</span>
    <span class="n">x</span><span class="o">=</span><span class="n">r_x</span><span class="p">,</span> 
    <span class="n">y</span><span class="o">=</span><span class="n">r_y</span><span class="p">,</span>
    <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span>
        <span class="n">size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
        <span class="n">line</span><span class="o">=</span><span class="n">Line</span><span class="p">(</span>
            <span class="n">color</span><span class="o">=</span><span class="s1">&#39;purple&#39;</span><span class="p">,</span>
            <span class="n">width</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
        <span class="n">opacity</span><span class="o">=</span><span class="mf">0.5</span><span class="p">),</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;breadth regression&quot;</span><span class="p">)</span>

<span class="n">layout</span> <span class="o">=</span> <span class="n">Layout</span><span class="p">(</span>
    <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Breadth of expression vs average expression level&quot;</span><span class="p">,</span>
    <span class="n">xaxis</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="s1">&#39;breadth&#39;</span><span class="p">,</span>
    <span class="p">),</span>
    <span class="n">yaxis</span><span class="o">=</span><span class="n">YAxis</span><span class="p">(</span>
        <span class="n">title</span><span class="o">=</span><span class="s1">&#39;average expression&#39;</span><span class="p">,</span>
    <span class="p">),</span>
<span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Data</span><span class="p">([</span><span class="n">trace1</span><span class="p">,</span> <span class="n">trace2</span><span class="p">]),</span> <span class="n">layout</span><span class="o">=</span><span class="n">layout</span><span class="p">)</span>
<span class="n">py</span><span class="o">.</span><span class="n">iplot</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt output_prompt">Out[52]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<iframe id="igraph" scrolling="no" style="border:none;"seamless="seamless" src="https://plot.ly/~oxana/255.embed" height="525" width="100%"></iframe>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="4.-Heatmap-of-gene-expression">4. Heatmap of gene expression<a class="anchor-link" href="#4.-Heatmap-of-gene-expression">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Heatmap is another great way to visualize big amounts of data. It allows to clearly see the outliers and explore the 
general clustering patterns. Are genes in different tissues, but the same donor expressed similarly or do the same tissues
from different donors tend to cluster together? Do brains of newborns and adults differ in gene expression patterns? 
Heatmaps of gene expression can give you good leads to questions like these.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>There is one catch with generating a heatmap for biological samples using plot.ly. Labels of the heatmap will actually
be coordinates on the x and y axis. For the plot to look less cluttered, I have removed the grid and set dtick to 1. Setting autotick to False also proved useful in order to see all the samples correctly labeled.</p>
<p>To improve readability, one often also needs to process samples names. In our data, as you probably noticed, sample names include everything: tissue name, annotation, donor, age. The name becomes long and impossible to display in a plot. Simple shortening will not work with plot.ly though, since the coordinates must be unique!</p>
<p>For this tutorial I've cheated a bit, by just adding an integer to each shortened name, I'm sure you can handle the string processing of your samples names on your own ;-)</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[53]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">pdist</span><span class="p">,</span> <span class="n">squareform</span>

<span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;breadth&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot_id&#39;</span><span class="p">,</span> <span class="s1">&#39;avg&#39;</span><span class="p">]]</span>
<span class="n">short_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">]</span>
<span class="n">short_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">short_cols</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">short_cols</span><span class="p">),</span><span class="mi">1</span><span class="p">)]</span>
<span class="n">data_dist</span> <span class="o">=</span> <span class="n">pdist</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">as_matrix</span><span class="p">()</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">Data</span><span class="p">([</span>
    <span class="n">Heatmap</span><span class="p">(</span>
        <span class="n">z</span><span class="o">=</span><span class="n">squareform</span><span class="p">(</span><span class="n">data_dist</span><span class="p">),</span> <span class="n">colorscale</span><span class="o">=</span><span class="s1">&#39;YIGnBu&#39;</span><span class="p">,</span>
        <span class="n">x</span><span class="o">=</span><span class="n">short_cols</span><span class="p">,</span>
        <span class="n">y</span><span class="o">=</span><span class="n">short_cols</span><span class="p">,</span>     <span class="c1"># y-axis labels</span>
    <span class="p">)</span>
<span class="p">])</span>

<span class="n">layout</span> <span class="o">=</span> <span class="n">Layout</span><span class="p">(</span>
    <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Transcription profiling of human brain samples&#39;</span><span class="p">,</span>
    <span class="n">autosize</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>
    <span class="n">margin</span><span class="o">=</span><span class="n">Margin</span><span class="p">(</span>
        <span class="n">l</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
        <span class="n">b</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span>
        <span class="n">pad</span><span class="o">=</span><span class="mi">4</span>
    <span class="p">),</span>
    <span class="n">xaxis</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span>
        <span class="n">showgrid</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="c1"># remove grid</span>
        <span class="n">autotick</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="c1"># custom ticks</span>
        <span class="n">dtick</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>        <span class="c1"># show 1 tick per day</span>
    <span class="p">),</span>
    <span class="n">yaxis</span><span class="o">=</span><span class="n">YAxis</span><span class="p">(</span>
        <span class="n">showgrid</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>   <span class="c1"># remove grid</span>
        <span class="n">autotick</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span>   <span class="c1"># custom ticks</span>
        <span class="n">dtick</span><span class="o">=</span><span class="mi">1</span>           <span class="c1"># show 1 tick per day</span>
    <span class="p">),</span>
<span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">layout</span><span class="o">=</span><span class="n">layout</span><span class="p">)</span>
<span class="n">py</span><span class="o">.</span><span class="n">iplot</span><span class="p">(</span><span class="n">fig</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">900</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">900</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt output_prompt">Out[53]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<iframe id="igraph" scrolling="no" style="border:none;"seamless="seamless" src="https://plot.ly/~oxana/256.embed?width=875&height=875" height="900" width="900"></iframe>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="5.-Network-of-gene-interactions">5. Network of gene interactions<a class="anchor-link" href="#5.-Network-of-gene-interactions">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now at some point in our biological investigations, we've got to dig deeper and look at concrete genes/proteins we found interesting.</p>
<p>If you go back to our first plot, you'll see that one of the points that stand out corresponds to Q16352(Alpha-internexin, AINX_HUMAN). This gene demonstrates both high level of expression in substantia nigra and the difference between adult and newborn samples is also significant. Which kind of makes sense, since this protein is involved in the morphogenesis of neurons. One of the ways to find out more about a protein is to look at it's interaction networks.</p>
<p>I've downloaded the interaction network in tab-separated format from a popular database <a href="http://string-db.org/">string-db.org</a>, so there is nothing novel in plotting it, we are merely reproducing the graph on their website, but, hopefully, you'll be able to use it for your future contributions to science!</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[54]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython2"><pre><span></span><span class="kn">import</span> <span class="nn">networkx</span> <span class="kn">as</span> <span class="nn">nx</span>

<span class="kn">import</span> <span class="nn">plotly.plotly</span> <span class="kn">as</span> <span class="nn">py</span>
<span class="kn">from</span> <span class="nn">plotly.graph_objs</span> <span class="kn">import</span> <span class="o">*</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">genfromtxt</span><span class="p">(</span><span class="s1">&#39;http://figshare.com/download/file/2088824&#39;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">usecols</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">14</span><span class="p">],</span>
                  <span class="n">dtype</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;S5&#39;</span><span class="p">,</span><span class="s1">&#39;S5&#39;</span><span class="p">,</span><span class="s1">&#39;f8&#39;</span><span class="p">])</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">names</span>

<span class="n">G</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="n">G</span><span class="o">.</span><span class="n">add_weighted_edges_from</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

<span class="n">pos</span><span class="o">=</span><span class="n">nx</span><span class="o">.</span><span class="n">spring_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>

<span class="n">edge_trace</span> <span class="o">=</span> <span class="n">Scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[],</span> <span class="n">y</span><span class="o">=</span><span class="p">[],</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;lines&#39;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="n">x0</span><span class="p">,</span> <span class="n">y0</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> 
    <span class="n">x1</span><span class="p">,</span> <span class="n">y1</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">edge</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> 
    <span class="n">edge_trace</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">x0</span><span class="p">,</span> <span class="n">x1</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>
    <span class="n">edge_trace</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">y0</span><span class="p">,</span> <span class="n">y1</span><span class="p">,</span> <span class="bp">None</span><span class="p">]</span>

<span class="n">node_trace</span> <span class="o">=</span> <span class="n">Scatter</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="p">[],</span> <span class="n">y</span><span class="o">=</span><span class="p">[],</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;markers+text&#39;</span><span class="p">,</span>
                     <span class="n">text</span><span class="o">=</span><span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">(),</span>
                     <span class="n">textposition</span><span class="o">=</span><span class="s1">&#39;top&#39;</span><span class="p">,</span>
                     <span class="n">marker</span><span class="o">=</span><span class="n">Marker</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">))</span>
<span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">G</span><span class="o">.</span><span class="n">nodes</span><span class="p">():</span>
    <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">pos</span><span class="p">[</span><span class="n">node</span><span class="p">]</span>
    <span class="n">node_trace</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
    <span class="n">node_trace</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
   
<span class="n">fig</span> <span class="o">=</span> <span class="n">Figure</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Data</span><span class="p">([</span><span class="n">edge_trace</span><span class="p">,</span> <span class="n">node_trace</span><span class="p">]),</span>
             <span class="n">layout</span><span class="o">=</span><span class="n">Layout</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s1">&#39;AINX_HUMAN interaction network&#39;</span><span class="p">,</span>
                           <span class="n">showlegend</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">xaxis</span><span class="o">=</span><span class="n">XAxis</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">showticklabels</span><span class="o">=</span><span class="bp">False</span><span class="p">),</span>
                           <span class="n">yaxis</span><span class="o">=</span><span class="n">YAxis</span><span class="p">(</span><span class="n">showgrid</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">zeroline</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">showticklabels</span><span class="o">=</span><span class="bp">False</span><span class="p">)))</span>

<span class="n">py</span><span class="o">.</span><span class="n">iplot</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">
<div class="prompt output_prompt">Out[54]:</div>


<div class="output_html rendered_html output_subarea output_execute_result">
<iframe id="igraph" scrolling="no" style="border:none;"seamless="seamless" src="https://plot.ly/~oxana/257.embed" height="525" width="100%"></iframe>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now you see our protein under it's gene name(INA) in the center of the graph.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Now, that's all, folks! I hope you enjoyed this intro to exploratory bioinformatics and got inspired to create beautiful interactive visualizations for your biological data.</p>

</div>
</div>
</div>
 

{% endraw %}